acceptance/tests/base/dsl/helpers/host_helpers_test.rb

```diff
@@ -0,0 +1,1377 @@
+    assert_equal hosts.size, unique_outputs.size
+  end
+
+  step "#on allows assertions to be used in the optional block" do
+    on hosts, %Q{echo "${RANDOM}:${RANDOM}"} do
+      assert_match /\d+:\d+/, stdout
+    end
+  end
+
+  step "#shell raises an exception when remote command fails" do
+    assert_raises(Beaker::Host::CommandFailure) do
+      shell "/bin/nonexistent-command"
+    end
+  end
+
+  step "#shell makes command output available via `.stdout` on success" do
+    output = shell(%Q{echo "echo via on"}).stdout
+    assert_equal "echo via on\n", output
+  end
+
+  step "#shell makes command error output available via `.stderr` on success" do
+    output = shell("/bin/nonexistent-command", :acceptable_exit_codes => [0, 127]).stderr
+    assert_match /No such file/, output
+  end
+
+  step "#shell makes exit status available via `.exit_code`" do
+    status = shell(%Q{echo "echo via on"}).exit_code
+    assert_equal 0, status
+  end
+
+  step "#shell with :acceptable_exit_codes will not fail for named exit codes" do
+    result = shell "/bin/nonexistent-command", :acceptable_exit_codes => [0, 127]
+    output = result.stderr
+    assert_match /No such file/, output
+    status = result.exit_code
+    assert_equal 127, status
+  end
+
+  step "#shell with :acceptable_exit_codes will fail for other exit codes" do
+    assert_raises(Beaker::Host::CommandFailure) do
+      shell %Q{echo "echo via on"}, :acceptable_exit_codes => [127]
+    end
+  end
+
+  step "#shell will pass :environment options to the remote host as ENV settings" do
+    result = shell "env", { :environment => { 'FOO' => 'bar' } }
+    output = result.stdout
+
+    assert_match /\nFOO=bar\n/, output
+  end
+
+  step "#shell allows assertions to be used in the optional block" do
+    shell %Q{echo "${RANDOM}:${RANDOM}"} do
+      assert_match /\d+:\d+/, stdout
+    end
+  end
+
+  step "#scp_to fails if the local file cannot be found" do
+    remote_tmpdir = tmpdir_on hosts.first
+    assert_raises IOError do
+      scp_to hosts.first, "/non/existent/file.txt", remote_tmpdir
+    end
+  end
+
+  if test_scp_error_on_close?
+    step "#scp_to fails if the remote path cannot be found" do
+      Dir.mktmpdir do |local_dir|
+        local_filename = File.join(local_dir, "testfile.txt")
+        File.open(local_filename, "w") do |local_file|
+          local_file.puts "contents"
+        end
+
+        # assert_raises Beaker::Host::CommandFailure do
+        assert_raises RuntimeError do
+          scp_to hosts.first, local_filename, "/non/existent/remote/file.txt"
+        end
+      end
+    end
+  end
+
+  step "#scp_to creates the file on the remote system" do
+    Dir.mktmpdir do |local_dir|
+      local_filename = File.join(local_dir, "testfile.txt")
+      File.open(local_filename, "w") do |local_file|
+        local_file.puts "contents"
+      end
+      remote_tmpdir = tmpdir_on hosts.first
+
+      scp_to hosts.first, local_filename, remote_tmpdir
```

Comment:
A number of the hosts methods only work when a single host is provided, and fail in varied ways when an array of hosts are provided. Some of them work in all cases. So, the suite is exercising both options.